src/hotspot/share/runtime/safepoint.cpp

```diff
@@ -954,12 +931,9 @@ void ThreadSafepointState::handle_polling_page_exception() {
     StackWatermarkSet::after_unwind(self);
 
     // Process pending operation
-    SafepointMechanism::process_if_requested(self);
-    // We have to wait if we are here because of a handshake for object deoptimization.
-    if (self->is_obj_deopt_suspend()) {
-      self->wait_for_object_deoptimization();
+    {
+      ThreadInVMfromJava tivm(self);
```

Comment:
This doesn't look good to me. We are in low-level safepoint/handshake related code, but we call a higher-level abstraction for thread-state transition management, just to get the side-effect of processing the safepoint/handshake correctly. To me this suggests we're missing an appropriate API entry point to do what needs to be done. I would rather see something like:
```
SafepointMechanism::process_if_requested(self);
self->handle_special_runtime_exit_condition(true /* check asyncs */);
```
(assuming that is the correct action of course).

Reply:
Ok, initially I thought of doing that but then decided to use transition wrappers. I'll change it. 
I think ideally we would like to use a transition wrapper in SafepointSynchronize::handle_polling_page_exception() but given the complexity of ThreadSafepointState::handle_polling_page_exception() we have to do things more manual unfortunately.